File: src/leetCode/c++/Solution232.cpp
//Stack pop(), empty(), pop(), peek(): O(1) O(n), 除了moveToOld()時需要O(n)時間
class MyQueue {//Solution232
private:
    stack<int> newStack;
    stack<int> oldStack;

    void moveToOld(){
        if(!oldStack.empty()) return;//等舊的stack都pop到為空才可加入, 否則順序會亂掉

        while(!newStack.empty()){
            oldStack.push(newStack.top());
            newStack.pop();
        }
    }

public:
    MyQueue() {
    }

    void push(int x) {
        newStack.push(x);
    }

    int pop() {
        moveToOld();
        int top = oldStack.top();
        oldStack.pop();
        return top;
    }

    int peek() {
        moveToOld();
        return oldStack.top();
    }

    bool empty() {
        return newStack.empty() && oldStack.empty();
    }

};
/*
newStack={3,2,1,0}
oldStack={}

當peek或pop時, moveToOld

newStack={}
oldStack={0,1,2,3}
*/

File: src/leetCode/java/Solution232.java
package leetCode.java;

import java.util.*;

//Stack pop(), empty(), pop(), peek(): O(1) O(n), 除了moveToOld()時需要O(n)時間
class MyQueue {//Solution232
    Stack<Integer> newSt;
    Stack<Integer> oldSt;

    public MyQueue() {
        newSt = new Stack<>();
        oldSt = new Stack<>();
    }

    public void push(int x) {
        newSt.push(x);
    }

    public int pop() {
        moveToOld();
        return oldSt.pop();
    }

    public int peek() {
        moveToOld();
        return oldSt.peek();
    }

    public boolean empty() {
        return newSt.isEmpty() && oldSt.isEmpty();
    }

    private void moveToOld(){
        if(!oldSt.isEmpty()) return;

        while(!newSt.isEmpty())
            oldSt.push(newSt.pop());//反向入列
    }
}